### PR TITLE
[chihiro] Round-2: width×FiLM×cosine-EMA composition at 512d

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Your PR #4 (4L/512d/8h, `pejudvyd`) proved that doubling model width from 256d→512d
yields a clean +4.3% improvement on every axis vs the merged baseline, with `volume_pressure`
winning most (14.37 vs 15.21). Two other independent wins — per-block FiLM geometry
conditioning (PR #8, frieren, −4.9% abupt at 256d) and progressive cosine EMA anneal
(PR #13, norman, −9.0% abupt at 256d) — are verified but pending rebase onto yi.

**Hypothesis:** these three gains are orthogonal and should compose. FiLM adds per-car
geometry specialization (targets surface tokens), cosine EMA improves late-training
checkpoint quality (targets epoch selection), and 512d adds volumetric capacity
(targets volume pressure). Stacking all three on the 512d base should push
`abupt_axis_mean` below 14, potentially toward 12–13.

**Expected gain from composition:**
- 256d baseline: 17.39 (merged PR #9)
- FiLM alone (256d): 16.53 (−4.9%)
- Cosine EMA alone (256d): 15.82 (−9.0%)
- Width alone (512d, your PR #4): 16.64 (−4.3%)
- Composition estimate: ~12.5–13.5 (linear combination of improvements, likely with
  some interaction)

## Instructions

Start from `yi` (already has your 512d code from PR #4 squash-merge). You need to
port two code changes from other student branches, then run on the 512d config.

### Step 1 — Port cosine EMA from `norman/round1-progressive-ema-decay`

Look at PR #13 / branch `norman/round1-progressive-ema-decay` for the implementation.
The key changes are:

**Add to `Config`:**
```python
ema_decay_start: float = 0.0      # initial EMA decay (0.0 = disabled, use fixed ema_decay)
ema_decay_end: float = 0.9999     # final EMA decay at end of training
```

**In the EMA class,** add a `set_decay(new_decay: float)` method that sets `self.decay = new_decay`.

**In the training loop** (after each optimizer step, where EMA is currently updated):
```python
if cfg.ema_decay_start > 0.0 and cfg.use_ema:
    progress = min(global_step / max_steps, 1.0)
    cos_val = (1 - math.cos(math.pi * progress)) / 2.0
    current_ema_decay = cfg.ema_decay_start + cos_val * (cfg.ema_decay_end - cfg.ema_decay_start)
    ema.set_decay(current_ema_decay)
```

Wire `max_steps = total_steps_across_all_epochs` (compute from dataset size, batch size, epochs at job start).

### Step 2 — Port FiLM from `frieren/round1-geometry-film-conditioning`

Look at PR #8 / branch `frieren/round1-geometry-film-conditioning` for the full
implementation. The key pieces:

**Add to `Config`:**
```python
use_film: bool = False
film_encoder_dim: int = 64
```

**Add classes:**
- `GeomEncoder`: mean-pooled MLP over surface points → single geometry token per case
- `FiLMLayer`: `to_gamma_beta = Linear(geom_dim, 2*hidden_dim)` applied as `h * (1 + gamma) + beta`

**In the training loop**, if `cfg.use_film`, encode the surface geometry ONCE per batch,
then modulate each Transolver block's hidden state after each attention layer.

Use AdaLN-zero init: initialize `to_gamma_beta.weight` and `to_gamma_beta.bias` to zero
so FiLM starts as identity (gamma=0, beta=0 → no modification). This is critical for
training stability with FiLM.

### Step 3 — Compose on 512d config

Run with ALL of the following:

```bash
cd target/
python train.py \
  --model-hidden-dim 512 --model-heads 8 --model-layers 4 --model-slices 128 \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay 0.9995 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --use-film \
  --use-tangential-wallshear-loss \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group b06-width-film-ema-composition
```

**Why `lr=5e-5`:** your PR #4 established that 512d diverges at `lr=2e-4`. Keep `5e-5`.
**Why `bs=4`:** largest batch fitting 96GB at 512d. Keep it.
**Why tangential projection:** already on yi (from PR #11 kohaku), adds the physics-aware
wall-shear constraint at zero extra cost. Adds no new code.

### Step 4 — Report

In your PR results comment, report:
1. Full `test_primary/*` table
2. Epoch-by-epoch val trajectory (confirm val improves monotonically — if best_epoch=1,
   flag it immediately, gradient clipping PR #22 may be needed)
3. VRAM peak usage and iteration speed (it/s)
4. Any gradient norm spikes visible in W&B

## Baseline to beat

Current merged yi baseline (PR #4 chihiro, W&B run `pejudvyd`):

| Metric | yi best (target to beat) | AB-UPT public |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **16.64** | — |
| `surface_pressure_rel_l2_pct` | 10.65 | 3.82 |
| `wall_shear_rel_l2_pct` | 17.66 | 7.29 |
| `volume_pressure_rel_l2_pct` | 14.37 | 6.08 |
| `wall_shear_x_rel_l2_pct` | 14.87 | 5.35 |
| `wall_shear_y_rel_l2_pct` | 19.89 | 3.65 |
| `wall_shear_z_rel_l2_pct` | 21.73 | 3.63 |

Lower is better. Your run already established the 512d baseline at 16.64 — your target
here is to get composition synergy below the sum-of-parts.

**Pending merges in flight** (will update yi soon):
- PR #8 frieren FiLM (16.53 at 256d) — rebasing now
- PR #13 norman cosine EMA (15.82 at 256d) — rebasing now

If these merge before you start, their code will be on yi and you can skip the manual
port above — just use the flags `--use-film` and `--ema-decay-start 0.99 --ema-decay-end 0.9999`
directly. Check `python train.py --help` to confirm flag availability before porting.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT chihiro:

## Results

**Composition LOST.** Test `abupt_axis_mean = 37.08` vs PR #4 baseline `16.64` (+123% worse). Root cause: the cosine EMA schedule horizon was set to 50 epochs but only ~3 epochs were achievable in the 6h budget, so EMA decay was effectively pinned at 0.99 — far weaker averaging than the baseline's fixed 0.9995. FiLM AdaLN-zero never ramped meaningfully off identity in only 3 epochs.

### `test_primary/*` table (W&B run `kk7wkhkv`)

| Metric | b06 (this run) | yi best (PR #4) | AB-UPT | Δ vs yi |
|---|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **37.08** | 16.64 | — | **+122.8%** |
| `surface_pressure_rel_l2_pct` | 11.50 | 10.65 | 3.82 | +8.0% |
| `wall_shear_rel_l2_pct` | 45.21 | 17.66 | 7.29 | +156.0% |
| `volume_pressure_rel_l2_pct` | 14.89 | 14.37 | 6.08 | +3.6% |
| `wall_shear_x_rel_l2_pct` | 31.39 | 14.87 | 5.35 | +111.1% |
| `wall_shear_y_rel_l2_pct` | 44.98 | 19.89 | 3.65 | +126.2% |
| `wall_shear_z_rel_l2_pct` | **82.66** | 21.73 | 3.63 | **+280.5%** |
| `surface_pressure_mae` | 0.0307 | — | — | — |
| `wall_shear_mae` | 0.4063 | — | — | — |
| `volume_pressure_mae` | 31.89 | — | — | — |

`full_val_primary/abupt = 36.39` (consistent with `test_primary`). Surface and volume held roughly even; the loss came almost entirely from wall shear, with `tau_z` essentially unrecoverable.

### Val trajectory

Validation ran after each epoch (`--validation-every 1`):

| Epoch | global_step | val_primary/abupt | sp | ws | vp | ws_z |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 21766 | 37.79 | 15.13 | 45.67 | 11.93 | 78.91 |
| 2 | 43532 | 36.67 | 11.62 | 45.42 | 9.33 | 85.86 |
| 3* | 43788 | **36.39** | 11.48 | 45.21 | 8.81 | 85.31 |

(*epoch 3 was a 256-step partial before timeout; `best_epoch=3`.) The headline `val_primary/abupt` is monotonically improving — no flat-on-epoch-1 pathology. Surface and volume pressure also improve cleanly across epochs (sp 15.13 → 11.62 → 11.48; vp 11.93 → 9.33 → 8.81). But **wall-shear-z gets *worse* across epochs** (78.91 → 85.86 → 85.31). FiLM-by-AdaLN-zero is identity at step 0, but the moment any gradient flows back through `to_gamma_beta`, FiLM begins expressing per-block multiplicative gating; in this run that signal apparently destabilises wall-shear-z faster than the rest of the model can compensate, and weak EMA (decay=0.99) bakes the drift directly into the eval shadow rather than averaging it out.

### Diagnosis: cosine EMA schedule horizon mismatch

`total_estimated_steps = max_epochs (50) × len(train_loader) (21766) = 1,088,300`. Training only completed `43,788` steps = **4.0% of the schedule**. The cosine ramp `(1 − cos(π × progress)) / 2` therefore moved from 0.0 → 0.0040, mapping to:

- Effective `train/ema_decay`: **0.9900 → 0.99004** (W&B confirms: min=0.990000, max=0.990039 over 43,789 logged values).
- Equivalent EMA averaging window: ~100 steps (vs ~2,000 steps at the baseline's fixed 0.9995).

At this stage of training the model is still rapidly improving — a 100-step window snaps too close to recent volatile updates and misses the better late-training state, while a 2000-step window catches it. The `tau_z` collapse is exactly what you'd expect from a mis-averaged EMA on the noisiest channel (wall-shear z had the highest variance of the seven outputs already at the baseline).

### Why FiLM didn't save it

AdaLN-zero init means `to_gamma_beta = (W=0, b=0)`, so FiLM is exact identity at step 0 (verified before launch: `with-geom vs no-geom max abs diff = 0.00e+00`). FiLM weights ramp off zero proportional to gradient × LR × steps; with only 3 epochs and `lr=5e-5`, the FiLM modulation is still small. Per-block FiLM thus contributed minimal lift at this training horizon while paying the EMA-schedule cost.

Grad norms: `P50=3.93, P95=19.75, P99=35.92, max=52.07 at step 2400`. Clipping wasn't enabled — these are spikes that an EMA at 0.9995 would smooth out but an EMA at 0.99 propagates almost directly to the eval shadow.

### Run details

- Command:
```bash
cd target/
python train.py \
  --model-hidden-dim 512 --model-heads 8 --model-layers 4 --model-slices 128 \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay 0.9995 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --use-film \
  --use-tangential-wallshear-loss \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group b06-width-film-ema-composition \
  --wandb-name "chihiro/b06-512d-film-ema" \
  --agent chihiro
```
- Model: 12.98M params (vs 12.7M PR #4; +280K from FiLM + GeomEncoder)
- Peak VRAM: **60.27 GB / 96 GB** (37% headroom; FiLM is cheap)
- Wall-clock: 285.1 min (timed out at SENPAI_TIMEOUT_MINUTES=360, val/test reserved 90 min)
- Iter speed: **2.56 it/s** average (≈3.3% slower than PR #4 baseline at 2.65 it/s — FiLM cost is small)
- Best epoch: 3 / 50; total steps: 43,788
- W&B: `wandb-applied-ai-team/senpai-v1-drivaerml/kk7wkhkv`

### Implementation notes

Ported FiLM from `frieren/round1-geometry-film-conditioning` (PR #8) and cosine EMA from `norman/round1-progressive-ema-decay` (PR #13). Both source PRs branched off pre-yi state, so I ported the deltas onto current yi (preserving tangential WS loss, clip_grad_norm, stochastic-depth code paths). New CLI flags exposed via dataclass: `--use-film`, `--film-encoder-dim`, `--ema-decay-start`, `--ema-decay-end`. Logged `train/ema_decay` per step. Pre-launch sanity: `with-geom == no-geom` exact at AdaLN-zero init (zero diff), perturbation produces expected non-zero diff.

A researcher-agent pass confirmed the design: FiLM AdaLN-zero + cosine EMA `0.99 → 0.9999` is the *right* mitigation for the zero-init shadow-lag concern (DiT precedent at static 0.9999); mask-aware mean-pool is correct vs max-pool for sign-bearing features; cosine LR + cosine EMA running full-duration are anti-correlated in the beneficial direction. The design was sound; the **schedule horizon assumption (50 epochs realistically achievable) was wrong for the 4L/512d/8h × 6h budget**.

### What happened — honest analysis

The hypothesis was that three orthogonal wins would compose. Two of the three (width, FiLM) were unable to express themselves in 3 epochs, and the third (cosine EMA) actively hurt us because its schedule horizon assumed 50 epochs and we did 3. Net: a strict regression vs the 512d width-only baseline.

The composition wasn't disproven — it was never tested. The cosine EMA schedule defaulted-into-bad behaviour the moment training was compute-bounded to <10% of the configured `--epochs`. PR #13's win at 256d/7L benefited from many more steps fitting in the same wall-clock; the same schedule at 4L/512d under the same wall-clock cap is essentially "fixed EMA at 0.99," which is much weaker than the merged baseline's 0.9995.

### Suggested follow-ups (in priority order)

1. **Make `total_estimated_steps` schedule-aware of the real budget.** Compute the cosine-schedule denominator from `(SENPAI_TIMEOUT_MINUTES × 60 × it_s_estimate)` or from a separate `--ema-schedule-steps` flag, not from `epochs × len(train_loader)`. Same fix should apply to the LR cosine annealing's `T_max`. This would turn the current run's "EMA ≈ 0.99 the whole time" into "EMA ramps 0.99 → ~0.999 across the actual 3 epochs," which is the intended design.

2. **Re-run the composition with the schedule fix.** With realistic horizon, the cosine EMA mid-training is ~0.997, comparable to but slightly below baseline 0.9995, and FiLM gets the same training time. This is the actual composition test.

3. **Isolate if (1)+(2) still loses:**
   - **b06b**: 512d + FiLM + fixed `ema_decay=0.9995` (no cosine). Tests "does FiLM transfer from 256d→512d at all?" PR #8 used 256d/7L with much more relative training; FiLM may need either a longer warmup or a non-trivial geometry token to show value at 512d/4L.
   - **b06c**: 512d + cosine EMA only (`--no-use-film`) with the schedule fix. Tests "does cosine EMA transfer from 7L→4L at the same total step budget?"

4. **Consider an LR-aware EMA schedule.** Anchor `ema_decay = 1 - lr/lr_max × (1 - 0.99)` so EMA naturally tightens as LR cools. Independent of `--epochs` and self-correcting under timeout.

5. **PR #22 grad clipping is likely beneficial regardless** — `P99=36, max=52` is high enough that clipping at `~25` would dampen the worst spikes without chopping median behavior. Probably worth pairing with this composition once the schedule fix is in.

I have not implemented any of these in this PR — flagging only.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #29 because ADVISOR: Closing — composition failed due to EMA schedule horizon mismatch (abupt=37.08, +122% over baseline). The cosine EMA schedule denominator was set to 50 epochs × steps/epoch, but only 3 epochs completed (4% of schedule), so EMA was effectively pinned at 0.99 instead of ramping to 0.9995+. This is a fundamental implementation bug, not a hypothesis failure.

The FiLM+512d+cosine-EMA composition remains a promising direction. With PR #13 (cosine EMA) now merged on yi — using actual-steps-elapsed as the schedule denominator — the next composition attempt should use the yi-native cosine EMA implementation. This PR is superseded by PR #22 (gradient clipping) + PR #13 (cosine EMA) + PR #14 (6L depth) which together represent the correct stack. Closing as the approach was broken by the schedule bug, not the hypothesis..
